lib/charms/acme_client_operator/v0/acme_client.py

```diff
@@ -0,0 +1,239 @@
+        _container = self.unit.get_container(self._container_name)
+        if not self.unit.is_leader():
+            return
+
+        if not _container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for container to be ready")
+            event.defer()
+            return
+
+        try:
+            csr = x509.load_pem_x509_csr(event.certificate_signing_request.encode())
+            subject_value = csr.subject.get_attributes_for_oid(NameOID.COMMON_NAME)[0].value
+            if isinstance(subject_value, bytes):
+                subject = subject_value.decode()
+            else:
+                subject = subject_value
+        except Exception:
+            logger.exception("Bad CSR received, aborting")
+            return
+
+        _container.push(
+            path=self._csr_path, make_dirs=True, source=event.certificate_signing_request.encode()
+        )
+
+        logger.info("Received Certificate Creation Request for domain %s", subject)
+        process = _container.exec(
+            self._cmd, timeout=300, working_dir="/tmp", environment=self._plugin_config
+        )
+        try:
+            stdout, error = process.wait_output()
+            logger.info(f"Return message: {stdout}, {error}")
+        except ExecError as e:
+            self.unit.status = BlockedStatus("Error getting certificate. Check logs for details")
+            logger.error("Exited with code %d. Stderr:", e.exit_code)
+            for line in e.stderr.splitlines():  # type: ignore
+                logger.error("    %s", line)
+            return
+
+        chain_pem = _container.pull(path=f"{self._certs_path}{subject}.crt")
+        certs = []
+        for cert in chain_pem.read().split("\n\n"):  # type: ignore[arg-type]
+            certs.append(cert)
+        self.tls_certificates.set_relation_certificate(
+            certificate=certs[0],
+            certificate_signing_request=event.certificate_signing_request,
+            ca=certs[-1],
+            chain=list(reversed(certs)),
+            relation_id=event.relation_id,
+        )
+
+    @property
+    def _cmd(self) -> List[str]:
+        """Command to run to get the certificate.
+
+        Returns:
+            list[str]: Command and args to run.
+        """
+        return [
+            "lego",
+            "--email",
+            self._email,
+            "--accept-tos",
+            "--csr",
+            self._csr_path,
+            "--server",
+            self._server,
+            "--dns",
+            self._plugin,
+            "run",
+        ]
+
+    @property
+    @abstractmethod
+    def _plugin_config(self) -> Dict[str, str]:
+        """Plugin specific additional configuration for the command.
+
+        Implement this method in your charm to return a dictionary with the plugin specific
+        configuration.
+
+        Returns:
+            dict[str, str]: Plugin specific configuration.
+        """
+
+    def _email_is_valid(self, email: str):
```

Comment:
I know that's in the lib but we should add return typing to some of these methods

Reply:
✅ 
Fixed on the lib side too, please take a look at the pr 
https://github.com/canonical/acme-client-base/pull/6